server/src/setup/http/express/versioned-api-router.ts

```diff
@@ -3,16 +3,14 @@ import { APIRouter } from './types'
 import { Controllers } from '../../application'
 import { UserRouter } from '../../../modules/users/infra/http/routes/user-router'
 
+
 const setupV1APIRouter = (controllers: Controllers): APIRouter => {
-  const endpointRouters = {
-    users: UserRouter.using(controllers.createUser),
-  }
 
   const router = Router()
   router.get('/', (_req, res) => {
     return res.json({ message: 'Water, water, water. Loo, loo loo.' })
   })
-  router.use('/users', endpointRouters.users)
+  router.use('/users/', UserRouter.using(controllers))
```

Comment:
A few days ago I fell down a rabbit hole of whether to end urls with a trailing `/` or not, and didn't end up coming to any conclusions in favour of either way (people seem to disagree both on which way to do it, but they also disagree on what the main reasons are for preferring their way). Do you have any reasons for prefering a trailing slash here? This is actually a pretty important thing to set a Loo Labs standard for.

Reply:
No specific preference, willing to lean either way.

Reply:
Dang, that makes both of us with no preference. @jeffzh4ng any thoughts?

Reply:
What did you find in the rabbit hole @xujustinj? I feel like the trailing slash indicates some sort of concatenation, and if the path is just `/users` then we should omit the trailing slash.

Reply:
### Colm Divilly's [blog](https://blog.cdivilly.com/2019/02/28/uri-trailing-slashes)

This is a tl;dr of https://cdivilly.wordpress.com/2014/03/11/why-trailing-slashes-on-uris-are-important/. The main argument is that trailing slashes make relative linking possible. Counterintuitively, relative links are not actually handled by concatenation.

### StackOverflow [thread](https://softwareengineering.stackexchange.com/questions/186959/trailing-slash-in-restful-api)

* REST has no opinion on this
* (grumbling about *nix file structures)
* trailing slash is better for relative linking
* reserve trailing slashes for "directories" (e.g. `/clubs/` is a directory of clubs, but `/clubs/123abc` has no trailing slash because it's a single resource)
* trailing slash indicates that a resource is _not_ a file, which helps SEO in the pathological case where the end of a url may be misinterpreted as a file extension

### Issue 118 in [pyeve/eve](https://github.com/pyeve/eve)

(not linking directly because that'll trigger a backlink to this discussion)

Arguments go both ways, but tl;dr: it's very commonplace to leave out the trailing slash, but it's technically wrong to do so.

### Vladimir Sidorenko's [article](https://gearheart.io/articles/restful-api-design-best-practices/)

> AVOID USING A FORWARD SLASH AT THE END OF THE URL
> A trailing forward slash (“/”) has no semantic meaning and can lead to confusion.

This seems to go against the relative linking argument, but I guess what they're getting at here is that relative linking has no use in a REST API.

### Guy Levin's [blog](https://blog.restcase.com/7-rules-for-rest-api-uri-design/)

> Rule #1: A trailing forward slash (/) should not be included in URIs
> This is one the most important rules to follow as the last character within a URI’s path, a forward slash (/) adds no semantic value and may cause confusion. REST API’s should not expect a trailing slash and should not include them in the links that they provide to clients.

He also points out that a lot of tools/frameworks treat trailing-slash and no-trailing-slash equivalently (by redirecting one to the other). This is why the question of trailing slashes is so invisible and barely thought about in the modern web.

Reply:
Thanks for the detailed right up. I do like the semantic reasoning of directories vs single resources on the SO thread.  But on the other hand you have other resources like pyeve/eve saying it's technically wrong to have trailing slashes. Feel free to make the final decision on this one @xujustinj @ShreyasPrasad